### ISSUE TITLE
Can the loss become significantly high at the beginning when using padding-free?

### ISSUE THREAD

**Issue by YooSungHyun** [huggingface/trl#2834]
### Reproduction

```python
from trl import ...
import argparse
import logging
import os

import numpy as np
import pandas as pd
import torch._dynamo
from accelerate import PartialState
from datasets import Dataset
from torch.utils.data.dataloader import _utils
from transformers import AutoTokenizer
from transformers.trainer_utils import is_main_process
from trl import (
    DataCollatorForCompletionOnlyLM,
    ModelConfig,
    ScriptArguments,
    SFTConfig,
    SFTTrainer,
    TrlParser,
    get_peft_config,
)
from utils import prepare_exp, read_txt, write_txt
from utils.Collator import preprocess_fft

_utils.MP_STATUS_CHECK_INTERVAL = 600.0
torch._dynamo.config.verbose = True
os.environ["TORCHDYNAMO_DISABLE"] = "1"
os.environ["TOKENIZERS_PARALLELISM"] = "false"

IGNORE_INDEX = -100


def load_train_dataset() -> Dataset:

    # private code
    # load excel and make df

    return human_dataset


def main(script_args: ScriptArguments, training_args: SFTConfig, model_args: ModelConfig) -> None:

    # (1) Preparing experiment
    prepare_exp("direct-packing", training_args)

    # (2) Load trainable model
    tokenizer = AutoTokenizer.from_pretrained(model_args.model_name_or_path, padding_side="right", use_fast=True)

    # (3) Load and preprocess train dataset
    with PartialState().local_main_process_first():
        train_dataset = load_train_dataset()
        train_dataset = train_dataset.map(
            preprocess_fft,
            fn_kwargs={
                "tokenizer": tokenizer,
                "system_prompt": raft_sys_prompt,
                "user_prompt_template": raft_user_prompt,
                "label_prompt": raft_label_prompt,
                "ignore_idx": IGNORE_INDEX,
            },
            remove_columns=list(train_dataset.features),
            num_proc=training_args.dataset_num_proc,
        )
        train_dataset = train_dataset.shuffle(seed=training_args.seed)
        train_dataset = train_dataset.filter(lambda x: len(x["input_ids"]) <= training_args.max_seq_length)

    training_args.dataset_kwargs = {"skip_prepare_dataset": True}
    qwen_instruction_template = "<|im_start|>user\n"
    qwen_response_template = "<|im_start|>assistant\n"
    llama_instruction_template = "<|start_header_id|>user<|end_header_id|>\n\n"
    llama_response_template = "<|start_header_id|>assistant<|end_header_id|>\n\n"
    collator = DataCollatorForCompletionOnlyLM(
        tokenizer=tokenizer,
        instruction_template=qwen_instruction_template,
        response_template=qwen_response_template,
        padding_free=True,
    )

    # (4) Run training
    trainer = SFTTrainer(
        model=model_args.model_name_or_path,
        args=training_args,
        train_dataset=train_dataset,
        processing_class=tokenizer,
        data_collator=collator,
        peft_config=get_peft_config(model_args),
    )
    trainer.train()

    # (5) Save outputs
    trainer.save_model(training_args.output_dir)
    if training_args.push_to_hub:
        trainer.push_to_hub(dataset_name=script_args.dataset_name)


def make_parser(subparsers: argparse._SubParsersAction = None):
    dataclass_types = (ScriptArguments, SFTConfig, ModelConfig)
    if subparsers is not None:
        parser = subparsers.add_parser("sft", help="Run the SFT training script", dataclass_types=dataclass_types)
    else:
        parser = TrlParser(dataclass_types)
    return parser


if __name__ == "__main__":
    parser = make_parser()
    script_args, training_args, model_args = parser.parse_args_and_config()
    logging.basicConfig(
        format="%(asctime)s - %(levelname)s - %(name)s -   %(message)s",
        datefmt="%m/%d/%Y %H:%M:%S",
        level=logging.INFO if is_main_process(training_args.local_rank) else logging.WARN,
    )
    main(script_args, training_args, model_args)

```

```bash
# model_name_or_path="unsloth/Meta-Llama-3.1-8B-Instruct"
model_name_or_path="Qwen/Qwen2.5-0.5B-Instruct"
exp_name="packing_llama"

export OMP_NUM_THREADS="16"
export TOKENIZERS_PARALLELISM="false"
# export NCCL_P2P_DISABLE=1


deepspeed --include localhost:0,1,2,3 --master_port 61000 train_fft.py \
    --output_dir=./outputs/${exp_name} \
    --dataset_name=none \
    --model_name_or_path=${model_name_or_path} \
    --learning_rate=3e-6 \
    --lr_scheduler_type=cosine \
    --per_device_train_batch_size=8 \
    --gradient_accumulation_steps=1 \
    --evaluation_strategy=no \
    --num_train_epochs=3 \
    --save_strategy=epoch \
    --logging_strategy=steps \
    --logging_steps=1 \
    --save_total_limit=1 \
    --remove_unused_columns=False \
    --dataloader_num_workers=32 \
    --warmup_ratio=0.03 \
    --gradient_checkpointing=True \
    --torch_compile=True \
    --neftune_noise_alpha=5 \
    --optim=adamw_torch \
    --report_to=none \
    --run_name=${exp_name} \
    --bf16 \
    --use_liger=True \
    --torch_dtype=bfloat16 \
    --max_seq_length=8192 \
    --attn_implementation=flash_attention_2 \
    --deepspeed=./config/zero2_llama_v0.6.json
```

outputs:
### packing use (batch 8 * 4gpu)

<img width="1527" alt="Image" src="https://github.com/user-attachments/assets/1824f768-5f85-4efa-bf7b-a50026827eb5" />

#### green line

<img width="1527" alt="Image" src="https://github.com/user-attachments/assets/7631c141-2069-46fb-8665-271b3c34d152" />

### packing unuse

<img width="1527" alt="Image" src="https://github.com/user-attachments/assets/c1810a33-7f75-4868-af8b-831e274da870" />

## my opinion
Please check the loss scale and grad_norm.  
Since the loss scale appears to be affected by batch size, it seems to be influenced by the length or number of data samples.

When training a typical instruct model, I did not observe significant differences in loss and grad_norm regardless of whether packing was used.  

However, my current dataset is an MRC task (providing a document and a question, then generating an answer based on the reference).  And I used zero2 not offload

Could this be due to the difference in datasets? Or did I make a mistake in data preprocessing?  
Is this phenomenon expected when using packing? Or could it be a bug in the Liger kernel?  

If anyone has experienced a similar issue, I would appreciate any insights.  

I observed the same trend with both `unsloth/llama3.1-8B-instruct` and `qwen/qwen2.5-0.5B-instruct`, so it does not seem to be related to the model itself.

### System Info

- Platform: Linux-5.15.0-60-generic-x86_64-with-glibc2.35
- Python version: 3.11.11
- PyTorch version: 2.5.1
- CUDA device(s): NVIDIA A100-SXM4-80GB, NVIDIA A100-SXM4-80GB, NVIDIA A100-SXM4-80GB, NVIDIA A100-SXM4-80GB, NVIDIA A100-SXM4-80GB, NVIDIA A100-SXM4-80GB, NVIDIA A100-SXM4-80GB, NVIDIA A100-SXM4-80GB
- Transformers version: 4.48.3
- Accelerate version: 1.3.0
- Accelerate config: not found
- Datasets version: 3.2.0
- HF Hub version: 0.27.1
- TRL version: 0.14.0
- bitsandbytes version: 0.45.1
- DeepSpeed version: 0.16.3
- Diffusers version: not installed
- Liger-Kernel version: 0.5.3
- LLM-Blender version: not installed
- OpenAI version: 1.60.2
- PEFT version: 0.14.0

### Checklist

- [x] I have checked that my issue isn't already filed (see [open issues](https://github.com/huggingface/trl/issues?q=is%3Aissue))
- [x] I have included my system information
- [x] Any code provided is minimal, complete, and reproducible ([more on MREs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any code provided is properly formatted in code blocks, (no screenshot, [more on code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any traceback provided is complete

**Comment by symoon11** [comment]
To the best of my knowledge, the "padding free" option works correctly only when FlashAttention is activated. It seems that FlashAttention is not currently activated. I recommend first creating the model and then passing it to the SFTTrainer.

**Comment by YooSungHyun** [comment]
@symoon11 thx for reply! i give `--attn_implementation=flash_attention_2`, but this is model_config and not training_args...
i will test soon and make some result

**Comment by YooSungHyun** [comment]
I foolishly forgot to include this part in my code:

```python
quantization_config = get_quantization_config(model_args)
model_kwargs = dict(
    revision=model_args.model_revision,
    trust_remote_code=model_args.trust_remote_code,
    attn_implementation=model_args.attn_implementation,
    torch_dtype=model_args.torch_dtype,
    use_cache=False if training_args.gradient_checkpointing else True,
    device_map=get_kbit_device_map() if quantization_config is not None else None,
    quantization_config=quantization_config,
)
training_args.model_init_kwargs = model_kwargs
```

Sorry for causing confusion. My code runs correctly now!
